diff --git a/WORKSPACE.bazel b/WORKSPACE.bazel
index a0f4c7bc1..ab6c04dab 100644
--- a/WORKSPACE.bazel
+++ b/WORKSPACE.bazel
@@ -46,8 +46,8 @@ http_archive(
 
 http_archive(
     name = "com_google_googleurl",
-    sha256 = "a1bc96169d34dcc1406ffb750deef3bc8718bd1f9069a2878838e1bd905de989",  # Last updated 2022-04-04
-    urls = ["https://storage.googleapis.com/quiche-envoy-integration/googleurl_9cdb1f4d1a365ebdbcbf179dadf7f8aa5ee802e7.tar.gz"],
+    sha256 = "3db90606ea9ca123d11f85c6fa3627a1e1e403d38c7fd1854a628cf1da3be5e2",  # Last updated 2022-04-04
+    urls = ["https://storage.googleapis.com/quiche-envoy-integration/googleurl-9cdb1f4d1a365ebdbcbf179dadf7f8aa5ee802e7.tar.gz"],
 )
 
 http_archive(
