diff --git a/src/tls/xqc_tls.c b/src/tls/xqc_tls.c
index 3b68679..07c840d 100644
--- a/src/tls/xqc_tls.c
+++ b/src/tls/xqc_tls.c
@@ -1061,7 +1061,9 @@ xqc_ssl_cert_cb(SSL *ssl, void *arg)
         return XQC_SSL_FAIL;
     }

-    hostname = SSL_get_servername(ssl, TLSEXT_NAMETYPE_host_name);
+    // hostname = SSL_get_servername(ssl, TLSEXT_NAMETYPE_host_name);
+    hostname = "test.xquic.com";
+
     if (NULL == hostname) {
         xqc_log(tls->log, XQC_LOG_ERROR, "|hostname is NULL");
         return XQC_SSL_FAIL;
diff --git a/tests/test_server.c b/tests/test_server.c
index 93a6706..389397e 100644
--- a/tests/test_server.c
+++ b/tests/test_server.c
@@ -2315,8 +2315,13 @@ int main(int argc, char *argv[]) {

     xqc_engine_ssl_config_t  engine_ssl_config;
     memset(&engine_ssl_config, 0, sizeof(engine_ssl_config));
-    engine_ssl_config.private_key_file = "./server.key";
-    engine_ssl_config.cert_file = "./server.crt";
+
+    // engine_ssl_config.private_key_file = "./server.key";
+    // engine_ssl_config.cert_file = "./server.crt";
+
+    engine_ssl_config.private_key_file = "./xquic.key";
+    engine_ssl_config.cert_file = "./xquic.crt";
+
     engine_ssl_config.ciphers = XQC_TLS_CIPHERS;
     engine_ssl_config.groups = XQC_TLS_GROUPS;

@@ -2617,6 +2622,7 @@ int main(int argc, char *argv[]) {
         return 0;
     }

+
     ctx.ev_socket = event_new(eb, ctx.fd, EV_READ | EV_PERSIST, xqc_server_socket_event_callback, &ctx);

     event_add(ctx.ev_socket, NULL);
