diff --git a/aflnet/afl.rs/src/bin/cargo-afl.rs b/aflnet/afl.rs/src/bin/cargo-afl.rs
index 66d9c73..3e55a1c 100644
--- a/aflnet/afl.rs/src/bin/cargo-afl.rs
+++ b/aflnet/afl.rs/src/bin/cargo-afl.rs
@@ -282,7 +282,8 @@ where
     I: IntoIterator<Item = S>,
     S: AsRef<OsStr>,
 {
-    let cargo_path = env!("CARGO");
+    // let cargo_path = env!("CARGO");
+    let cargo_path = "cargo";
 
     // add some flags to sanitizers to make them work with Rust code
     let asan_options = env::var("ASAN_OPTIONS").unwrap_or_default();
