diff --git a/picoquic/quicctx.c b/picoquic/quicctx.c
index bc8bd2cf..d7aef267 100644
--- a/picoquic/quicctx.c
+++ b/picoquic/quicctx.c
@@ -1456,13 +1456,31 @@ int picoquic_get_version_index(uint32_t proposed_version)
 
 static void picoquic_create_random_cnx_id(picoquic_quic_t* quic, picoquic_connection_id_t * cnx_id, uint8_t id_length)
 {
+    // QUIC-Fuzz ADDED HERE
+    // hardcode connection ID
+    static int conn_id = 1;
+
     if (id_length > 0) {
-        picoquic_crypto_random(quic, cnx_id->id, id_length);
+        // picoquic_crypto_random(quic, cnx_id->id, id_length);
+
+        for(int i=0; i< id_length; i++){
+            cnx_id->id[i] = conn_id;
+        }
+
+        conn_id += 1;
+
+        if(conn_id == 255){
+            conn_id = 1;
+        }
     }
+
     if (id_length < sizeof(cnx_id->id)) {
         memset(cnx_id->id + id_length, 0, sizeof(cnx_id->id) - id_length);
     }
+
     cnx_id->id_len = id_length;
+
+    // QUIC-Fuzz ADDED END
 }
 
 void picoquic_create_local_cnx_id(picoquic_quic_t* quic, picoquic_connection_id_t* cnx_id, uint8_t id_length, picoquic_connection_id_t cnx_id_remote)
diff --git a/picoquic/sockloop.c b/picoquic/sockloop.c
index 7fc038f4..755546d2 100644
--- a/picoquic/sockloop.c
+++ b/picoquic/sockloop.c
@@ -81,6 +81,8 @@
 
 #include <pthread.h>
 
+int pthread_setname_np(pthread_t thread, const char *name);  // Explicit declaration
+
 #ifndef SOCKET_TYPE
 #define SOCKET_TYPE int
 #endif
diff --git a/picoquicfirst/picoquicdemo.c b/picoquicfirst/picoquicdemo.c
index b5b331be..306e70c0 100644
--- a/picoquicfirst/picoquicdemo.c
+++ b/picoquicfirst/picoquicdemo.c
@@ -345,6 +345,12 @@ int quic_server(const char* server_name, picoquic_quic_config_t * config, int ju
         }
     }
 
+// QUIC-Fuzz ADDED HERE
+// #ifdef __AFL_HAVE_MANUAL_CONTROL
+//   __AFL_INIT();
+// #endif
+// QUIC-Fuzz ADDED END
+
     if (ret == 0) {
         /* Wait for packets */
 #if _WINDOWS_BUT_WE_ARE_UNIFYING
